### PR TITLE
Log gl error redo

### PR DESCRIPTION
**Pull request by Kovak**
This PR reverses the commits made in #3420 while implementing the same logc with a new structure

A discussion on IRC with matham yielded decisions to follow a different structure, the key changes are:

1. Add a config.pxi declaration 'HIDE_GL_LOG' that will control whether or not the gl logging works. This option defaults to False as we think that the added overhead is fairly small compared to the benefit of providing users with more in depth information when a gl error occurs. 

2. Have the inline function declared in a .pxi for easy inlining everywhere, originally common.pxi was considered but do to conflicts in the opengl.pyx a new gl_debug_logger.pxi was introduced.

I have added the new function to several of the existing graphics classes in this PR as well, but the work here is incomplete as I'm still hunting down gl invocations.

### CODE REVIEW

File: kivy/graphics/gl_debug_logger.pxi [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from kivy.graphics.c_opengl cimport GLenum, glGetError
+from kivy.logger import Logger
+include "config.pxi"
+
+cdef inline void log_gl_error(str note):
+    IF not HIDE_GL_LOG:
```
Comment (tito) [inline kivy/graphics/gl_debug_logger.pxi]: Does this code works with HIDE_GL_LOG? If the IF is evaluate to False, the result will be an empty function, not even an pass.

**Comment by tito** [comment]
Instead of having `HIDE_GL_LOG`, could we use `DEBUG_GL` = True instead? default to False.

File: kivy/graphics/gl_debug_logger.pxi (hunk 1 above)
Comment (Kovak) [inline kivy/graphics/gl_debug_logger.pxi]: Yes if it is False we get:

```c
static CYTHON_INLINE void __pyx_f_4kivy_8graphics_6shader_log_gl_error(CYTHON_UNUSED PyObject *__pyx_v_note) {
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("log_gl_error", 0);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
}
```

which is pretty much an empty function

if True:

```c
static CYTHON_INLINE void __pyx_f_4kivy_8graphics_6shader_log_gl_error(PyObject *__pyx_v_note) {
  GLenum __pyx_v_ret;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  PyObject *__pyx_t_7 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("log_gl_error", 0);

  /* "kivy\graphics\gl_debug_logger.pxi":7
 * cdef inline void log_gl_error(str note):
 *     IF not HIDE_GL_LOG:
 *         ret = glGetError()             # <<<<<<<<<<<<<<
 *         if ret:
 *             Logger.error("OpenGL Error: {note} {ret1} / {ret2}".format(
 */
  __pyx_v_ret = glGetError();

  /* "kivy\graphics\gl_debug_logger.pxi":8
 *     IF not HIDE_GL_LOG:
 *         ret = glGetError()
 *         if ret:             # <<<<<<<<<<<<<<
 *             Logger.error("OpenGL Error: {note} {ret1} / {ret2}".format(
 *                 note=note, ret1=ret, ret2=hex(ret)))
 */
  __pyx_t_1 = (__pyx_v_ret != 0);
  if (__pyx_t_1) {

    /* "kivy\graphics\gl_debug_logger.pxi":9
 *         ret = glGetError()
 *         if ret:
 *             Logger.error("OpenGL Error: {note} {ret1} / {ret2}".format(             # <<<<<<<<<<<<<<
 *                 note=note, ret1=ret, ret2=hex(ret)))
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_Logger); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_error); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_OpenGL_Error_note_ret1_ret2, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = PyDict_New(); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);

    /* "kivy\graphics\gl_debug_logger.pxi":10
 *         if ret:
 *             Logger.error("OpenGL Error: {note} {ret1} / {ret2}".format(
 *                 note=note, ret1=ret, ret2=hex(ret)))             # <<<<<<<<<<<<<<
 */
    if (PyDict_SetItem(__pyx_t_5, __pyx_n_s_note, __pyx_v_note) < 0) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_6 = __Pyx_PyInt_From_GLenum(__pyx_v_ret); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 10; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    if (PyDict_SetItem(__pyx_t_5, __pyx_n_s_ret1, __pyx_t_6) < 0) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = __Pyx_PyInt_From_GLenum(__pyx_v_ret); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 10; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_7 = PyTuple_New(1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 10; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_6);
    __Pyx_GIVEREF(__pyx_t_6);
    __pyx_t_6 = 0;
    __pyx_t_6 = __Pyx_PyObject_Call(__pyx_builtin_hex, __pyx_t_7, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 10; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    if (PyDict_SetItem(__pyx_t_5, __pyx_n_s_ret2, __pyx_t_6) < 0) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

    /* "kivy\graphics\gl_debug_logger.pxi":9
 *         ret = glGetError()
 *         if ret:
 *             Logger.error("OpenGL Error: {note} {ret1} / {ret2}".format(             # <<<<<<<<<<<<<<
 *                 note=note, ret1=ret, ret2=hex(ret)))
 */
    __pyx_t_6 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_empty_tuple, __pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_4, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_6); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_GOTREF(__pyx_t_2);
    } else {
      __pyx_t_3 = PyTuple_New(1+1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_3, 0+1, __pyx_t_6);
      __Pyx_GIVEREF(__pyx_t_6);
      __pyx_t_6 = 0;
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_3, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    goto __pyx_L3;
  }
  __pyx_L3:;

  /* "kivy\graphics\gl_debug_logger.pxi":5
 * include "config.pxi"
 * 
 * cdef inline void log_gl_error(str note):             # <<<<<<<<<<<<<<
 *     IF not HIDE_GL_LOG:
 *         ret = glGetError()
 */

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_7);
  __Pyx_WriteUnraisable("kivy.graphics.shader.log_gl_error", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
}
```

Comment (Kovak) [inline kivy/graphics/gl_debug_logger.pxi]: Adding a pass

```pyx
cdef inline void log_gl_error(str note):
    IF DEBUG_GL:
        ret = glGetError()
        if ret:
            Logger.error("OpenGL Error: {note} {ret1} / {ret2}".format(
                note=note, ret1=ret, ret2=hex(ret)))
    ELSE:
        pass
```

results in generating the same code:
```c
static CYTHON_INLINE void __pyx_f_4kivy_8graphics_6shader_log_gl_error(CYTHON_UNUSED PyObject *__pyx_v_note) {
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("log_gl_error", 0);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
}
```

**Comment by kived** [comment]
Make sure this is tested with kivy-ios. It should probably work better using a .pxi for the inline function though: #3426

**Comment by skylite21** [comment]
I did a successful compile with Python3.4.1 on OSX 10.10.3 with this branch (Cython==0.21.2)

**Pull request merged**